hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/MessageFlyweight.java

```diff
@@ -175,7 +175,11 @@ public String getStringUtf8() {
     }
 
     public Data getData() {
-        return new HeapData(getByteArray());
+        final int length = buffer.getInt(index + offset);
+        index += Bits.INT_SIZE_IN_BYTES;
+        Data data = new ByteBufferHeapData(buffer.byteArray(), index + offset, length);
```

Comment:
Ok. I did not see this coming. I am avoiding copy but it causes more memory usage where these are stored. 
So this improvement will be beneficial everywhere Data comes as response to client. When client message going to remote, it is not beneficial since it needs to be stored on cluster. 
What I can do is to keep the improvement but actually identify the places where Data needs to be stored. And copy to a smaller buffer only needed. 
For the data coming to client, only place seems like the places we use caching. 
How does that sound ?